torchao/prototype/low_bit_optim/README.md

```diff
@@ -46,6 +46,54 @@ lpmm 4-bit (*) |  7.73           | 11m 10s                  | 89.71
 
 (*) means rank-1 normalization is used for 2nd optimizer state. Refer to [paper](https://arxiv.org/abs/2309.01507) for more details.
 
+## Optimizer CPU offload
+
+This folder also implements optimizer CPU offload (i.e. ZeRO-Offload) for single GPU training. For multi-GPU training, you can use FSDP's built-in CPU offload.
+
+```python
+import torch
+from torchao.prototype.low_bit_optim import CPUOffloadOptimizer
+
+model = ...
+
+# only offload optimizer state
+optim = CPUOffloadOptimizer(model.parameters(), torch.optim.AdamW, fused=True)
+
+# offload optimizer state AND gradients
+optim = CPUOffloadOptimizer(model.parameters(), torch.optim.AdamW, offload_gradients=True, fused=True)
+```
+
+This will reduce GPU memory usage by optimizer state size, and additionally gradient size if `offload_gradients=True`. `CPUOffloadOptimizer` can wrap any base optimizer.
+
+NOTE:
+- Since the optimizer step is done on CPU, it is highly recommended to use a fast CPU optimizer, such as `torch.optim.AdamW(fused=True)` (requires PyTorch 2.4). For other optimizers, you can try `torch.compile()` their optimizer step.
+- To minimize the amount of CPU<->GPU data transfer, we keep a copy of parameters and pre-allocate gradients memory on CPU. Therefore, expect your RAM usage to increase by 2x model size + optimizer state (which is 2x model size for Adam).
+- It is recommended NOT to `torch.compile()` your whole model when `CPUOffloadOptimizer` is used, as it prevents us from interleaving gradient device-to-host transfer with backward pass. To minimize such impact, you can compile parts of your model separately.
```

Comment:
it's not clear to me from the test or benchmark when this specific point is relevant

Reply:
Are you referring to the one about torch.compile? (sorry not sure which line you are referring to from GitHub UI). If it is, I can add benchmark for this config in README, and also its trace in the PR description (to compare with eager).